subsys/bluetooth/controller/ll_sw/ull_adv_sync.c

```diff
@@ -195,66 +197,131 @@ uint8_t ll_adv_sync_enable(uint8_t handle, uint8_t enable)
 		 * random address.
 		 */
 	} else {
-		sync->is_enabled = 1U;
+		sync_got_enabled = 1U;
 	}
 
 	if (adv->is_enabled && !sync->is_started) {
-		volatile uint32_t ret_cb = TICKER_STATUS_BUSY;
-		uint32_t ticks_anc_sync;
-		uint8_t err;
+		uint32_t ticks_slot_overhead_aux;
+		struct lll_adv_aux *lll_aux;
+		struct ll_adv_aux_set *aux;
+		uint32_t ticks_anchor_sync;
+		uint32_t ticks_anchor_aux;
+		uint32_t ret;
 
-		/* FIXME: Find absolute ticks until after auxliary PDU on air
-		 *        to place the periodic advertising PDU.
-		 */
-		ticks_anc_sync = ticker_ticks_now_get();
+		lll_aux = adv->lll.aux;
 
 		/* Add sync_info into auxiliary PDU */
 		err = ull_adv_aux_hdr_set_clear(adv,
 						ULL_ADV_PDU_HDR_FIELD_SYNC_INFO,
-						0, NULL, NULL);
+						0, NULL, NULL, &pri_idx);
 		if (err) {
 			return err;
 		}
 
-		ull_hdr_init(&sync->ull);
+		if (lll_aux) {
+			/* FIXME: Find absolute ticks until after auxiliary PDU
+			 *        on air to place the periodic advertising PDU.
+			 */
+			ticks_anchor_aux = 0U; /* unused in this path */
+			ticks_slot_overhead_aux = 0U; /* unused in this path */
+			ticks_anchor_sync = ticker_ticks_now_get();
+			aux = NULL;
+		} else {
+			lll_aux = adv->lll.aux;
+			aux = (void *)HDR_LLL2EVT(lll_aux);
+			ticks_anchor_aux = ticker_ticks_now_get();
+			ticks_slot_overhead_aux = ull_adv_aux_evt_init(aux);
+			ticks_anchor_sync =
+				ticks_anchor_aux + ticks_slot_overhead_aux +
+				aux->evt.ticks_slot +
+				HAL_TICKER_US_TO_TICKS(EVENT_MAFS_US);
+		}
 
-		err = ull_adv_sync_start(sync, ticks_anc_sync, &ret_cb);
-		if (err) {
-			return BT_HCI_ERR_CMD_DISALLOWED;
+		ret = ull_adv_sync_start(sync, ticks_anchor_sync);
+		if (ret) {
+			err_val = BT_HCI_ERR_INSUFFICIENT_RESOURCES;
+			goto sync_cleanup;
 		}
 
 		sync->is_started = 1U;
+
+		lll_adv_data_enqueue(&adv->lll, pri_idx);
+
+		if (aux) {
+			/* Keep aux interval equal or higher than primary PDU
+			 * interval.
+			 */
+			aux->interval = adv->interval +
+					(HAL_TICKER_TICKS_TO_US(
+						ULL_ADV_RANDOM_DELAY) /	625U);
+
+			ret = ull_adv_aux_start(aux, ticks_anchor_aux,
+						ticks_slot_overhead_aux);
+			if (ret) {
+				err_val = BT_HCI_ERR_INSUFFICIENT_RESOURCES;
+				goto sync_cleanup;
+			}
+
+			aux->is_started = 1U;
+		}
+	}
+
+	if (sync_got_enabled) {
+		sync->is_enabled = sync_got_enabled;
 	}
 
 	return 0;
+
+sync_cleanup:
+	/* Remove sync_info from auxiliary PDU */
+	err = ull_adv_aux_hdr_set_clear(adv, 0,
+					ULL_ADV_PDU_HDR_FIELD_SYNC_INFO,
+					NULL, NULL, &pri_idx);
+	if (err) {
+		if (err_val) {
+			return err_val;
+		}
+
+		return err;
+	}
+
+	lll_adv_data_enqueue(&adv->lll, pri_idx);
+
+	if (sync->is_started) {
+		/* TODO: we removed sync info, but if sync_stop() fails, what do
+		 * we do?
+		 */
+		err = sync_stop(sync);
+		if (err) {
+			if (err_val) {
+				return err_val;
+			}
+
+			return err;
+		}
+
+		sync->is_started = 0U;
+	}
+
+	if (!enable) {
+		sync->is_enabled = 0U;
+	}
+
+	return err_val;
```

Comment:
Please move this into a separate function. I do not think the potential savings warrant the use of goto in this case.

Reply:
Moved it to an inline function.